### PR TITLE
Validation tests for queue.writeBuffer

### PR DESCRIPTION
**Pull request by haoxli**
- The ranges tests have covered these validations:
  - the write size is  too large for the destination buffer,
  - the write size extends the source buffer size.
- Add tests for buffer offset and data offset out of range.
- Add tests for the buffer missed COPY_DST usage.
- Cleanup TODOs.

<hr>

**Author checklist for test code/plans:**

- [x] All outstanding work is tracked with "TODO" in a test/file description or `.unimplemented()` on a test.
- [x] New helpers, if any, are documented using `/** doc comments */` and can be found via `helper_index.txt`.
- [x] (Optional, sometimes not possible.) Tests pass (or partially pass without unexpected issues) in an implementation. (Add any extra details above.)

**[Reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) for test code/plans:** (Note: feel free to pull in other reviewers at any time for any reason.)

- [x] The test path is reasonable, the [description](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) "describes the test, succinctly, but in enough detail that a reader can read only the test plans in a file or directory and evaluate the completeness of the test coverage."
- [ ] Tests appear to cover this area completely, except for outstanding TODOs. Validation tests use control cases.
    (This is critical for coverage. Assume anything without a TODO will be forgotten about forever.)
- [x] Existing (or new) test helpers are used where they would reduce complexity.
- [x] TypeScript code is readable and understandable (is unobtrusive, has reasonable type-safety/verbosity/dynamicity).

### CODE REVIEW

File: src/webgpu/api/validation/queue/writeBuffer.spec.ts [hunk 1]
```diff
@@ -1,21 +1,12 @@
...
-- dataOffset {ok, too large for data}
-- buffer size {ok, too small for copy}
-- data size {ok, too small for copy}
-- size {aligned, unaligned}
-- size unspecified; default {ok, too large for buffer}
-
 Note: destroyed buffer is tested in destroyed/.
-
-TODO: implement usage flag validation.
-TODO: validate large write sizes that may overflow.
```
Comment (kainino0x) [inline src/webgpu/api/validation/queue/writeBuffer.spec.ts:15]: This TODO is about `size` or `offset+size` overflowing a 64-bit integer, and I don't think we test it yet, so please leave the TODO (can add clarification) - or add the test cases in this PR.

File: src/webgpu/api/validation/queue/writeBuffer.spec.ts [hunk 2]
```diff
@@ -109,16 +100,21 @@ Also verifies that the specified data range:
...
+      // Writing with a data offset that is out of range of data size
+      t.shouldThrow('OperationError', () => queue.writeBuffer(buffer, 0, arraySm, 9, 0));
+
       // A data offset of undefined should be treated as 0
       queue.writeBuffer(buffer, 0, arraySm, undefined, 8);
       t.shouldThrow('OperationError', () => queue.writeBuffer(buffer, 0, arraySm, undefined, 12));
     }
 
     const arrayTypes = [
-      Uint8Array,
```
[the remark this answers is not in the document]
Comment (haoxli) [reply src/webgpu/api/validation/queue/writeBuffer.spec.ts:121]: Maybe, only `Uint8ClampedArray` is missed.

File: src/webgpu/api/validation/queue/writeBuffer.spec.ts (hunk 1 above)
Comment (haoxli) [reply src/webgpu/api/validation/queue/writeBuffer.spec.ts:15]: The maximum safe value of Number is 2^54 - 1, we cannot make `size` or `data size`  in bytes overflow a 64-bit integer, the value larger than Number.MAX_SAFE_INTEGER are not acceptable.

Comment (kainino0x) [reply src/webgpu/api/validation/queue/writeBuffer.spec.ts:15]: That's the largest safe value, which means that every integer can be represented. After that, the next representable value after `x` is `x+2`, `x+4`, ..., `y`, `y+4`, `y+8`..., and so on. However you can still have values larger than MAX_SAFE_INTEGER that can be converted to uint64 (`unsigned long long`) by WebIDL, I believe.

The largest JS number (double) less than 2^64 appears to be `0xfffffffffffff800`, which should get converted to uint64, and then you could add an offset like `0x800` to it, and it would overflow to 0.

File: src/webgpu/api/validation/queue/writeBuffer.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kainino0x) [reply src/webgpu/api/validation/queue/writeBuffer.spec.ts:121]: If the tests pass with Uint8ClampedArray, let's add it now.

File: src/webgpu/api/validation/queue/writeBuffer.spec.ts [hunk 3]
```diff
@@ -132,6 +133,29 @@ Also verifies that the specified data range:
...
+g.test('write_sizes_overflow')
+  .desc(
+    `
+Tests writing large sizes that may overflow a 64-bit integer.
+`
+  )
+  .paramsSubcasesOnly([
+    { size: 16, _valid: true }, // control case
+    { size: 0xfffffffffffff800, _valid: false }, // The largest number less than the maximum of uint64
+    { size: 0xffffffffffffff00, _valid: false }, // The smallest number overflow uint64
```
Comment (kainino0x) [inline src/webgpu/api/validation/queue/writeBuffer.spec.ts:145]: Sorry I wasn't clear about what I meant by "add an offset". These are useful to test WebIDL implementations, but they should already be robust to this kind of overflow. The WebGPU-specific implementation (e.g. Blink or Dawn), however, is likely to do something like `offset + size` and we want to test that the implementation doesn't allow that to overflow. So cases are also needed which set `offset` to a larger value.

Comment (haoxli) [reply src/webgpu/api/validation/queue/writeBuffer.spec.ts]: Thanks for your clarification. 
As we discussed offline, I had filed an issue for the TypeError thrown if a Number larger that MAX_SAFE_INTEGER (https://bugs.chromium.org/p/dawn/issues/detail?id=961), and still expected it throw OperationError in our cases. If it's actually the WebIDL-required behavior, we will remove  the overflow case at that time.

Comment (kainino0x) [reply src/webgpu/api/validation/queue/writeBuffer.spec.ts]: We got an answer on the bug, and indeed it turns out it's invalid in WebIDL! Sorry for my misreading of the WebIDL spec. Let's remove the overflow cases then 👍

**Review by kainino0x — APPROVED** [review]

File: src/webgpu/api/validation/queue/writeBuffer.spec.ts (hunk 3 above)
Comment (haoxli) [reply src/webgpu/api/validation/queue/writeBuffer.spec.ts]: Got it! Removed.
